Add support for RecognitionContext (#76)
cobaltspeech/sdk-cubic: docs-src/content/using-cubic-sdk/recognition-context.md

---
title: "Recognition Context"
weight: 25
---

Cubic allows users to send context information with a recognition request which
may aid the speech recognition. For example, if you have a list of names that
you want to make sure the Cubic model transcribes correctly, with the correct
spelling, then you may provide the list in the form of a
[`RecognitionContext`](../../protobuf/autogen-doc-cubic-proto/#message-recognitioncontext) object along with the [`RecognitionConfig`](../client-configs/#fields)
before streaming data.

Cubic models allow different sets of "context tokens" each of which can be
paired with a list of words or phrases. For example, a Cubic model may have a
context token for airport names, and you can provide a list of airport names you
want to be recognized correctly for this context token. Likewise, models may
also be configured with tokens for "contact list names", "menu items", "medical jargon" etc.

To ensure that there is no added latency in processing the list of words or
phrases during a recognition request, we have a API method called
[`CompileContext()`](../../protobuf/autogen-doc-cubic-proto/#service-cubic) that

> allows the user to compile the list into a compact, fast to search format for
> the Cubic model. Currently, we only accept list of phrases or words that have
> been precompiled.

Reviewer: ```suggestion
allows the user to compile the list into a compact, efficient format for
passing to the `Recognize()` or `StreamingRecognize()` methods.
```

    "strings"

    "github.com/cobaltspeech/sdk-cubic/grpc/go-cubic"
    "github.com/cobaltspeech/sdk-cubic/grpc/go-cubic/cubicpb"
)

func main() {

    // creating client without TLS. Remove WithInsecure() if using TLS
    serverAddr := "127.0.0.1:2727"
    client, err := cubic.NewClient(serverAddr, cubic.WithInsecure())
    if err != nil {
        log.Fatal(err)
    }
    defer client.Close()

    // Get list of available models
    modelResp, err := client.ListModels(context.Background())
    if err != nil {
        log.Fatal(err)
    }
    for _, m := range modelResp.Models {
        fmt.Printf("\nID = %v, Name = %v, Supports Context = %v\n",
            m.Id, m.Name, m.Attributes.SupportsContext)
        // printing allowed context tokens
        if m.Attributes.SupportsContext {
            fmt.Printf("Allowed Context Tokens = %v\n\n",
                strings.Join(m.Attributes.AllowedContextTokens, ", "))
        }
    }

    // Assuming the first model supports context
    model := modelResp.Models[0]

    // Let's say this model has an allowed context token called "airport_names" and
    // we have a list of airport names that we want to make sure the recognizer gets

>     // right. We can compile the list of names into a easy to search and decodable
>     // format for the model using the CompileContext() method. We can save the compiled
>     // data and send it back during a recognize request to aid the speech recognition.

Reviewer: ```suggestion
    // right. We compile the list of names using the CompileContext(), save the compiled
    // data and send it back with subsequent recognize requests to customize and improve the results.
```

    // a small example list
    phrases := []string{"NARITA", "KUALA LUMPUR INTERNATIONAL", "ISTANBUL ATATURK", "LAGUARDIA"}
    contextToken := model.Attributes.AllowedContextTokens[0] // "airport_names"

    // sending request to server
    compiledResp, err := client.CompileContext(
        context.Background(), model.Id, contextToken, phrases, nil)
    if err != nil {
        log.Fatal(err)
    }

    // saving the compiled result for later use; note this compiled data is only
    // compatible with the model whose ID was provided in the CompileContext call
    compiledContexts := make([]*cubicpb.CompiledContext, 0)
    compiledContexts = append(compiledContexts, compiledResp.Context)

    // Now we can send a recognize request along with the compiled context. The
    // context data is provided through the recognition config as a list of compiled

>     // contexts (i.e we can provide more than one compiled context if the model

Reviewer: ```suggestion
    // contexts (i.e. we can provide more than one compiled context if the model
```

    resultHandler := func(resp *cubicpb.RecognitionResponse) {
        for _, r := range resp.Results {
            if !r.IsPartial {
                fmt.Println(r.Alternatives[0].Transcript)
            }
        }
    }

    err = client.StreamingRecognize(context.Background(), cfg, f, resultHandler)
    if err != nil {
        log.Fatal(err)
    }
}
{{< /tab >}}

{{< tab "Python" "python" >}}
import cubic

# set insecure to False if server uses TLS
serverAddress = '127.0.0.1:2727'
client = cubic.Client(serverAddress, insecure=True)

# Get list of available models
modelResp = client.ListModels()
for m in modelResp.models:
    print("\nID = {}, Name = {}, Supports Context = {}".format(
        m.id, m.name, m.attributes.supports_context))
    if m.attributes.supports_context:
        # printing allowed context tokens
        print("Allowed context tokens =  {}\n".format(
            str(m.attributes.allowed_context_tokens)))

# Assuming the first model supports context
model = modelResp.models[0]

# Let's say this model has an allowed context token called "airport_names" and
# we have a list of airport names that we want to make sure the recognizer gets

> # right. We can compile the list of names into a easy to search and decodable
> # format for the model using the CompileContext() method. We can save the compiled
> # data and send it back during a recognize request to aid the speech recognition.

Reviewer: ```suggestion
# right. We compile the list of names using the CompileContext(), save the compiled
# data and send it back with subsequent recognize requests to customize and improve the results.
```

# a small example list
phrases = ["NARITA", "KUALA LUMPUR INTERNATIONAL", "ISTANBUL ATATURK", "LAGUARDIA"]
contextToken = model.attributes.allowed_context_tokens[0]  # "airport_names"

# sending request to server
compiledResp = client.CompileContext(model.id, contextToken, phrases)

# saving the compiled result for later use; note this compiled data is only
# compatible with the model whose ID was provided in the CompileContext call
compiledContexts = []
compiledContexts.append(compiledResp.context)

# Now we can send a recognize request along with the compiled context. The
# context data is provided through the recognition config as a list of compiled

> # contexts (i.e we can provide more than one compiled context if the model

Reviewer: ```suggestion
# contexts (i.e. we can provide more than one compiled context if the model
```

            // Initialize a gRPC connection
            var channel = new Grpc.Core.Channel(serverAddress, creds);
            var client = new CobaltSpeech.Cubic.Cubic.CubicClient(channel);

            // Get list of available models
            var listModelsRequest = new CobaltSpeech.Cubic.ListModelsRequest();
            var modelResp = client.ListModels(listModelsRequest);

            foreach (var m in modelResp.Models) {
                Console.WriteLine("\nID = {0}, Name = {1}, Supports Context = {2}",
                    m.Id, m.Name, m.Attributes.SupportsContext);
                // printing allowed context tokens
                if (m.Attributes.SupportsContext) {
                    Console.WriteLine("Allowed Context Tokens = {0}\n",
                        m.Attributes.AllowedContextTokens);
                }
            }

            // Assuming the first model supports context
            var model = modelResp.Models[0];

            // Let's say this model has an allowed context token called "airport_names" and
            // we have a list of airport names that we want to make sure the recognizer gets

>             // right. We can compile the list of names into a easy to search and decodable
>             // format for the model using the CompileContext() method. We can save the compiled
>             // data and send it back during a recognize request to aid the speech recognition.

Reviewer: ```suggestion
            // right. We compile the list of names using the CompileContext(), save the compiled
            // data and send it back with subsequent recognize requests to customize and improve the results.
```

            // create compile context request
            var compileRequest = new CobaltSpeech.Cubic.CompileContextRequest {
                ModelId = model.Id,
                Token = contextToken,
            };
            // put phrases into an compileRequest.Phrases
            foreach (var phrase in phrases) {
                compileRequest.Phrases.Add(
                    new CobaltSpeech.Cubic.ContextPhrase {
                        Text = phrase,
                    });
            }

            // send request to server
            var compiledResp = client.CompileContext(compileRequest);

            // saving the compiled result for later use; note this compiled data is only
            // compatible with the model whose ID was provided in the CompileContext call
            var compiledContexts = new List<CobaltSpeech.Cubic.CompiledContext>();
            compiledContexts.Add(compiledResp.Context);

            // Now we can send a recognize request along with the compiled context. The
            // context data is provided through the recognition config as a list of compiled

>             // contexts (i.e we can provide more than one compiled context if the model

Reviewer: ```suggestion
            // contexts (i.e. we can provide more than one compiled context if the model
```

        dispatchGroup.enter()
        
        var model: Cobaltspeech_Cubic_Model!

        client.listModels(success: { (models) in
            guard let models = models else { return }
            
            for model in models {
                print("\nID = \(model.id), Name = \(model.name), Supports Context = \(model.attributes.supportsContext)")
                // printing allowed context tokens
                if model.attributes.supportsContext {
                    print("Allowed Context Tokens = \(model.attributes.allowedContextTokens)\n")
                }
            }
            
            // Assuming the first model supports context
            if let firstModel = models.first {
                model = firstModel
            } else {
                return
            }
            
            dispatchGroup.leave()
        }) { (error) in
            print(error.localizedDescription)
            dispatchGroup.leave()
        }
        
        dispatchGroup.wait()
        
        // Let's say this model has an allowed context token called "airport_names" and
        // we have a list of airport names that we want to make sure the recognizer gets

>         // right. We can compile the list of names into a easy to search and decodable
>         // format for the model using the CompileContext() method. We can save the compiled
>         // data and send it back during a recognize request to aid the speech recognition.

Reviewer: ```suggestion
        // right. We compile the list of names using the CompileContext(), save the compiled
        // data and send it back with subsequent recognize requests to customize and improve the results.
```

        for phrase in phrases {
            var contextPhrase = Cobaltspeech_Cubic_ContextPhrase()
            contextPhrase.text = phrase
            compileRequest.phrases.append(contextPhrase)
        }

        // send request to server
        var compiledContexts: [Cobaltspeech_Cubic_CompiledContext] = []
        
        dispatchGroup.enter()
        
        self.client.compileContext(compileRequest).response.whenComplete { (result) in
            switch result {
            case .success(let response):
                // saving the compiled result for later use; note this compiled data is only
                // compatible with the model whose ID was provided in the CompileContext call
                compiledContexts.append(response.context)
                dispatchGroup.leave()
            case .failure(let error):
                print(error.localizedDescription)
                dispatchGroup.leave()
            }
        }
        
        dispatchGroup.wait()
        
        // Now we can send a recognize request along with the compiled context. The
        // context data is provided through the recognition config as a list of compiled

>         // contexts (i.e we can provide more than one compiled context if the model

Reviewer: ```suggestion
        // contexts (i.e. we can provide more than one compiled context if the model
```

{{% /tab %}}

{{% tab "Swift/iOS" %}}

> ```swift

Author: Yes, will change will rebasing on to master
